File: Frontend/scripts/start.js
#!/usr/bin/env node

/**
 * SCRIPT DE ARRANQUE COMPLETO PARA FRONTEND SHEILY AI
 * =============================================================================
 * Este script verifica todas las dependencias, configura el entorno
 * y inicia el frontend Next.js en el puerto 3000 sin errores
 * =============================================================================
 * Ubicación: Frontend/scripts/start.js
 * =============================================================================
 */

const { spawn, execSync } = require('child_process');
const fs = require('fs');
const path = require('path');
const os = require('os');

// Colores para output
const Colors = {
    RED: '\x1b[31m',
    GREEN: '\x1b[32m',
    YELLOW: '\x1b[33m',
    BLUE: '\x1b[34m',
    PURPLE: '\x1b[35m',
    CYAN: '\x1b[36m',
    NC: '\x1b[0m'
};

// Función para imprimir mensajes con colores
function printStatus(message) {
    console.log(`${Colors.BLUE}[INFO]${Colors.NC} ${message}`);
}

function printSuccess(message) {
    console.log(`${Colors.GREEN}[SUCCESS]${Colors.NC} ${message}`);
}

function printWarning(message) {
    console.log(`${Colors.YELLOW}[WARNING]${Colors.NC} ${message}`);
}

function printError(message) {
    console.log(`${Colors.RED}[ERROR]${Colors.NC} ${message}`);
}

function printHeader(title) {
    console.log(`${Colors.PURPLE}${'='.repeat(32)}${Colors.NC}`);
    console.log(`${Colors.PURPLE}${title}${Colors.NC}`);
    console.log(`${Colors.PURPLE}${'='.repeat(32)}${Colors.NC}`);
}

class FrontendStarter {
    constructor() {
        this.projectDir = __dirname;
        this.packageJsonPath = path.join(this.projectDir, 'package.json');
        this.nodeModulesPath = path.join(this.projectDir, 'node_modules');
        this.envLocalPath = path.join(this.projectDir, '.env.local');
        this.port = 3000;
        this.hostname = '127.0.0.1';
    }

    // Función para ejecutar comandos de forma síncrona
    execCommand(command, options = {}) {
        try {
            return execSync(command, { 
                cwd: this.projectDir, 
                encoding: 'utf8',
                stdio: options.silent ? 'pipe' : 'inherit',
                ...options 
            });
        } catch (error) {
            if (!options.ignoreErrors) {
                throw error;
            }
            return null;
        }
    }

    // Función para ejecutar comandos de forma asíncrona
    spawnCommand(command, args, options = {}) {
        return new Promise((resolve, reject) => {
            const child = spawn(command, args, {
                cwd: this.projectDir,
                stdio: 'inherit',
                ...options
            });

            child.on('close', (code) => {
                if (code === 0) {
                    resolve();
                } else {
                    reject(new Error(`Command failed with exit code ${code}`));
                }
            });

            child.on('error', reject);
        });
    }

    // Limpiar procesos anteriores
    cleanupPreviousProcesses() {
        printStatus('Limpiando procesos anteriores del frontend...');
        
        try {
            // Buscar y matar procesos en el puerto 3000
            if (os.platform() === 'win32') {
                // Windows
                try {
                    const result = this.execCommand(`netstat -ano | findstr :${this.port}`, { silent: true });
                    if (result) {
                        printWarning('Encontrados procesos anteriores en puerto 3000');
                        const lines = result.split('\n');
                        for (const line of lines) {
                            if (line.trim()) {
                                const parts = line.trim().split(/\s+/);
                                if (parts.length >= 5) {
                                    const pid = parts[parts.length - 1];
                                    try {
                                        this.execCommand(`taskkill /PID ${pid} /F`, { silent: true, ignoreErrors: true });
                                    } catch (e) {
                                        // Ignorar errores
                                    }
                                }
                            }
                        }
                    }
                } catch (e) {
                    // Ignorar errores
                }
            } else {
                // Linux/Mac
                try {
                    const result = this.execCommand(`lsof -ti:${this.port}`, { silent: true });
                    if (result && result.trim()) {
                        const pids = result.trim().split('\n');
                        printWarning(`Encontrados procesos anteriores en puerto ${this.port}: ${pids.join(', ')}`);
                        for (const pid of pids) {
                            try {
                                this.execCommand(`kill -9 ${pid}`, { silent: true, ignoreErrors: true });
                            } catch (e) {
                                // Ignorar errores
                            }
                        }
                    }
                } catch (e) {
                    // Ignorar errores
                }
            }

            // Limpiar caché de Next.js
            const nextCachePath = path.join(this.projectDir, '.next');
            if (fs.existsSync(nextCachePath)) {
                printStatus('Limpiando caché de Next.js...');
                const cacheDir = path.join(nextCachePath, '.cache');
                const traceDir = path.join(nextCachePath, 'trace');
                
                if (fs.existsSync(cacheDir)) {
                    fs.rmSync(cacheDir, { recursive: true, force: true });
                }
                if (fs.existsSync(traceDir)) {
                    fs.rmSync(traceDir, { recursive: true, force: true });
                }
                printSuccess('Caché limpiado');
            }

            printSuccess('Procesos anteriores terminados');
        } catch (error) {
            printWarning(`No se pudieron limpiar procesos anteriores: ${error.message}`);
        }
    }

    // Verificar dependencias del sistema
    checkSystemDependencies() {
        printHeader('VERIFICANDO DEPENDENCIAS DEL SISTEMA');
        
        try {
            // Verificar Node.js
            const nodeVersion = this.execCommand('node --version', { silent: true });
            if (!nodeVersion) {
                printError('Node.js no está instalado');
                printStatus('Por favor, instala Node.js 18 o superior desde https://nodejs.org/');
                return false;
            }

            // Verificar npm
            const npmVersion = this.execCommand('npm --version', { silent: true });
            if (!npmVersion) {
                printError('npm no está instalado');
                return false;
            }

            printSuccess(`Node.js encontrado: ${nodeVersion.trim()}`);
            printSuccess(`npm encontrado: ${npmVersion.trim()}`);

            // Verificar versión mínima de Node.js
            const nodeMajor = parseInt(nodeVersion.trim().substring(1).split('.')[0]);
            if (nodeMajor < 18) {
                printError(`Se requiere Node.js 18 o superior. Versión actual: ${nodeVersion.trim()}`);
                return false;
            }

            printSuccess('Todas las dependencias del sistema están disponibles');
            return true;
        } catch (error) {
            printError(`Error al verificar dependencias: ${error.message}`);
            return false;
        }
    }

    // Configurar entorno
    setupEnvironment() {
        printHeader('CONFIGURANDO ENTORNO');
        
        try {
            // Verificar si estamos en el directorio correcto
            if (!fs.existsSync(this.packageJsonPath)) {
                printError('No se encontró package.json. Asegúrate de estar en el directorio Frontend/');
                return false;
            }

            // Verificar si es el proyecto correcto
            const packageData = JSON.parse(fs.readFileSync(this.packageJsonPath, 'utf8'));
            if (packageData.name !== 'sheily-landing-next') {
                printError('Este no parece ser el proyecto Sheily AI Frontend');
                return false;
            }

            printSuccess('Directorio del proyecto verificado');

            // Crear archivo .env.local si no existe
            if (!fs.existsSync(this.envLocalPath)) {
                printStatus('Creando archivo .env.local...');
                const envContent = `# Configuración del Frontend Sheily AI
NODE_ENV=development
PORT=${this.port}
HOSTNAME=${this.hostname}
NEXTAUTH_SECRET=sheily_ai_frontend_secret_key_development_${Date.now()}
BACKEND_URL=http://localhost:8000
NEXTAUTH_URL=http://${this.hostname}:${this.port}
`;
                fs.writeFileSync(this.envLocalPath, envContent);
                printSuccess('Archivo .env.local creado');
            } else {
                printSuccess('Archivo .env.local ya existe');
            }

            // Configurar variables de entorno
            process.env.NODE_ENV = 'development';
            process.env.PORT = this.port.toString();
            process.env.HOSTNAME = this.hostname;

            printSuccess('Entorno configurado correctamente');
            return true;
        } catch (error) {
            printError(`Error al configurar entorno: ${error.message}`);
            return false;
        }
    }

    // Verificar dependencias de Node.js
    checkNodeDependencies() {
        printHeader('VERIFICANDO DEPENDENCIAS DE NODE.JS');
        
        try {
            if (!fs.existsSync(this.nodeModulesPath)) {
                printWarning('node_modules no encontrado. Instalando dependencias...');
                return this.installDependencies();
            }

            printSuccess('node_modules encontrado');

            // Verificar dependencias
            try {
                const result = this.execCommand('npm ls --depth=0', { silent: true });
                if (result && (result.includes('UNMET DEPENDENCY') || result.includes('npm ERR!'))) {
                    printWarning('Dependencias faltantes detectadas. Reinstalando...');
                    return this.installDependencies();
                } else {
                    printSuccess('Todas las dependencias están instaladas correctamente');
                    return true;
                }
            } catch (error) {
                printWarning('Error al verificar dependencias. Reinstalando...');
                return this.installDependencies();
            }
        } catch (error) {
            printError(`Error al verificar dependencias: ${error.message}`);
            return false;
        }
    }

    // Instalar dependencias
    installDependencies() {
        printHeader('INSTALANDO DEPENDENCIAS');
        
        try {
            // Limpiar instalación anterior si existe
            if (fs.existsSync(this.nodeModulesPath)) {
                printStatus('Limpiando instalación anterior...');
                fs.rmSync(this.nodeModulesPath, { recursive: true, force: true });
                
                const packageLockPath = path.join(this.projectDir, 'package-lock.json');
                if (fs.existsSync(packageLockPath)) {
                    fs.unlinkSync(packageLockPath);
                }
            }

            // Limpiar caché de npm
            printStatus('Limpiando caché de npm...');
            this.execCommand('npm cache clean --force', { silent: true });

            // Instalar dependencias
            printStatus('Instalando dependencias...');
            this.execCommand('npm install');

            printSuccess('Dependencias instaladas correctamente');
            return true;
        } catch (error) {
            printError(`Error al instalar dependencias: ${error.message}`);
            return false;
        }
    }

    // Verificar configuración de Next.js
    checkNextJSConfig() {
        printHeader('VERIFICANDO CONFIGURACIÓN DE NEXT.JS');
        
        const configFiles = [
            'next.config.cjs',
            'tsconfig.json',
            'tailwind.config.ts',
            'postcss.config.cjs'
        ];

        for (const configFile of configFiles) {
            const configPath = path.join(this.projectDir, configFile);
            if (fs.existsSync(configPath)) {
                printSuccess(`${configFile} encontrado`);
            } else {
                printError(`${configFile} no encontrado`);
                return false;
            }
        }

        // Verificar configuración de TypeScript
        try {
            printStatus('Verificando configuración de TypeScript...');
            this.execCommand('npx tsc --noEmit --skipLibCheck', { silent: true, ignoreErrors: true });
            printSuccess('Configuración de TypeScript verificada');
        } catch (error) {
            printWarning('Errores de TypeScript detectados, pero continuando...');
        }

        printSuccess('Configuración de Next.js verificada');
        return true;
    }

    // Verificar archivos de audio
    checkAudioFiles() {
        printHeader('VERIFICANDO ARCHIVOS DE AUDIO');
        
        const audioDir = path.join(this.projectDir, 'public', 'sounds');
        const requiredFiles = [
            '062708_laser-charging-81968.mp3',
            'whoosh-drum-hits-169007.mp3'
        ];

        if (!fs.existsSync(audioDir)) {
            printWarning('Directorio de sonidos no encontrado. Creando...');
            fs.mkdirSync(audioDir, { recursive: true });
        }

        const missingFiles = [];
        for (const audioFile of requiredFiles) {
            if (!fs.existsSync(path.join(audioDir, audioFile))) {
                missingFiles.push(audioFile);
            }
        }

        if (missingFiles.length > 0) {
            printWarning(`Archivos de audio faltantes: ${missingFiles.join(', ')}`);
            printStatus('Los sonidos se generarán sintéticamente como fallback');
        } else {
            printSuccess('Todos los archivos de audio están disponibles');
        }

        return true;
    }

    // Verificar puertos
    checkPorts() {
        printHeader('VERIFICANDO PUERTOS');
        
        try {
            // Verificar puerto 3000
            if (os.platform() === 'win32') {
                const result = this.execCommand(`netstat -ano | findstr :${this.port}`, { silent: true, ignoreErrors: true });
                if (result && result.trim()) {
                    printWarning(`Puerto ${this.port} está en uso`);
                    printStatus('Intentando liberar el puerto...');
                    this.cleanupPreviousProcesses();
                }
            } else {
                const result = this.execCommand(`lsof -ti:${this.port}`, { silent: true, ignoreErrors: true });
                if (result && result.trim()) {
                    printWarning(`Puerto ${this.port} está en uso`);
                    printStatus('Intentando liberar el puerto...');
                    this.cleanupPreviousProcesses();
                }
            }

            printSuccess(`Puerto ${this.port} está disponible`);

            // Verificar puerto 8000 (backend)
            if (os.platform() === 'win32') {
                const result = this.execCommand('netstat -ano | findstr :8000', { silent: true, ignoreErrors: true });
                if (result && result.trim()) {
                    printSuccess('Backend detectado en puerto 8000');
                } else {
                    printWarning('Backend no detectado en puerto 8000');
                }
            } else {
                const result = this.execCommand('lsof -ti:8000', { silent: true, ignoreErrors: true });
                if (result && result.trim()) {
                    printSuccess('Backend detectado en puerto 8000');
                } else {
                    printWarning('Backend no detectado en puerto 8000');
                }
            }

            printStatus('El frontend funcionará en modo standalone si es necesario');
            return true;
        } catch (error) {
            printWarning(`No se pudieron verificar los puertos: ${error.message}`);
            return true; // Continuar de todas formas
        }
    }

    // Iniciar frontend
    async startFrontend() {
        printHeader('INICIANDO FRONTEND SHEILY AI');
        
        printStatus(`Iniciando servidor de desarrollo en puerto ${this.port}...`);
        printStatus(`URL: http://${this.hostname}:${this.port}`);
        printStatus('Presiona Ctrl+C para detener el servidor');
        
        try {
            // Iniciar el servidor
            const env = { ...process.env };
            env.NODE_ENV = 'development';
            env.PORT = this.port.toString();
            env.HOSTNAME = this.hostname;

            await this.spawnCommand('npm', ['run', 'dev'], { env });
        } catch (error) {
            if (error.message.includes('exit code')) {
                printStatus('Servidor detenido');
            } else {
                printError(`Error al iniciar el frontend: ${error.message}`);
            }
        }
    }

    // Ejecutar proceso completo
    async run() {
        printHeader('ARRANQUE DEL FRONTEND SHEILY AI');
        printStatus('Iniciando proceso de arranque completo...');
        
        try {
            // Ejecutar todas las verificaciones y configuraciones
            if (!this.checkSystemDependencies()) {
                return false;
            }

            if (!this.setupEnvironment()) {
                return false;
            }

            if (!this.checkNodeDependencies()) {
                return false;
            }

            if (!this.checkNextJSConfig()) {
                return false;
            }

            if (!this.checkAudioFiles()) {
                return false;
            }

            if (!this.checkPorts()) {
                return false;
            }

            printHeader('VERIFICACIONES COMPLETADAS');
            printSuccess('✅ Sistema de dependencias verificado');
            printSuccess('✅ Entorno configurado');
            printSuccess('✅ Configuración de Next.js validada');
            printSuccess('✅ Puertos verificados');
            printSuccess('✅ Archivos de audio verificados');

            printStatus('Iniciando frontend...');
            await this.startFrontend();
            return true;
        } catch (error) {
            printError(`Error durante el arranque: ${error.message}`);
            return false;
        }
    }
}

// Función principal
async function main() {
    try {
        const starter = new FrontendStarter();
        const success = await starter.run();
        
        if (!success) {
            printError('El arranque del frontend falló');
            process.exit(1);
        }
    } catch (error) {
        printError(`Error inesperado: ${error.message}`);
        process.exit(1);
    }
}

// Manejar señales de interrupción
process.on('SIGINT', () => {
    printStatus('Recibida señal de interrupción. Limpiando...');
    process.exit(0);
});

process.on('SIGTERM', () => {
    printStatus('Recibida señal de terminación. Limpiando...');
    process.exit(0);
});

// Ejecutar función principal
main();

File: Frontend/scripts/start.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
SCRIPT DE ARRANQUE COMPLETO PARA FRONTEND SHEILY AI
=============================================================================
Este script verifica todas las dependencias, configura el entorno
y inicia el frontend Next.js en el puerto 3000 sin errores
=============================================================================
Ubicación: Frontend/scripts/start.py
=============================================================================
"""

import os
import sys
import subprocess
import shutil
import json
import time
import signal
import platform
from pathlib import Path
from typing import List, Dict, Optional, Tuple


# Colores para output
class Colors:
    RED = "\033[0;31m"
    GREEN = "\033[0;32m"
    YELLOW = "\033[1;33m"
    BLUE = "\033[0;34m"
    PURPLE = "\033[0;35m"
    CYAN = "\033[0;36m"
    NC = "\033[0m"  # No Color


def print_status(message: str) -> None:
    """Imprime un mensaje de estado."""
    print(f"{Colors.BLUE}[INFO]{Colors.NC} {message}")


def print_success(message: str) -> None:
    """Imprime un mensaje de éxito."""
    print(f"{Colors.GREEN}[SUCCESS]{Colors.NC} {message}")


def print_warning(message: str) -> None:
    """Imprime un mensaje de advertencia."""
    print(f"{Colors.YELLOW}[WARNING]{Colors.NC} {message}")


def print_error(message: str) -> None:
    """Imprime un mensaje de error."""
    print(f"{Colors.RED}[ERROR]{Colors.NC} {message}")


def print_header(title: str) -> None:
    """Imprime un encabezado."""
    print(f"{Colors.PURPLE}{'='*32}{Colors.NC}")
    print(f"{Colors.PURPLE}{title}{Colors.NC}")
    print(f"{Colors.PURPLE}{'='*32}{Colors.NC}")


class FrontendStarter:
    """Clase principal para el arranque del frontend."""

    def __init__(self):
        self.project_dir = Path(__file__).parent
        self.package_json_path = self.project_dir / "package.json"
        self.node_modules_path = self.project_dir / "node_modules"
        self.env_local_path = self.project_dir / ".env.local"
        self.port = 3000
        self.hostname = "127.0.0.1"

    def cleanup_previous_processes(self) -> None:
        """Limpia procesos anteriores del frontend."""
        print_status("Limpiando procesos anteriores del frontend...")

        try:
            # Buscar procesos en el puerto 3000
            if platform.system() == "Windows":
                cmd = f"netstat -ano | findstr :{self.port}"
                result = subprocess.run(cmd, shell=True, capture_output=True, text=True)
                if result.stdout:
                    print_warning("Encontrados procesos anteriores en puerto 3000")
                    # Terminar procesos en Windows
                    for line in result.stdout.split("\n"):
                        if line.strip():
                            parts = line.split()
                            if len(parts) >= 5:
                                pid = parts[-1]
                                try:
                                    subprocess.run(
                                        f"taskkill /PID {pid} /F", shell=True
                                    )
                                except:
                                    pass
            else:
                # Linux/Mac
                cmd = f"lsof -ti:{self.port}"
                result = subprocess.run(cmd, shell=True, capture_output=True, text=True)
                if result.stdout.strip():
                    pids = result.stdout.strip().split("\n")
                    print_warning(
                        f"Encontrados procesos anteriores en puerto {self.port}: {pids}"
                    )
                    for pid in pids:
                        try:
                            subprocess.run(f"kill -9 {pid}", shell=True)
                        except:
                            pass
                    time.sleep(2)

            # Limpiar caché de Next.js
            next_cache = self.project_dir / ".next"
            if next_cache.exists():
                print_status("Limpiando caché de Next.js...")
                cache_dir = next_cache / ".cache"
                trace_dir = next_cache / "trace"
                if cache_dir.exists():
                    shutil.rmtree(cache_dir)
                if trace_dir.exists():
                    shutil.rmtree(trace_dir)
                print_success("Caché limpiado")

            print_success("Procesos anteriores terminados")

        except Exception as e:
            print_warning(f"No se pudieron limpiar procesos anteriores: {e}")

    def check_system_dependencies(self) -> bool:
        """Verifica las dependencias del sistema."""
        print_header("VERIFICANDO DEPENDENCIAS DEL SISTEMA")

        # Verificar Node.js
        if not shutil.which("node"):
            print_error("Node.js no está instalado")
            print_status(
                "Por favor, instala Node.js 18 o superior desde https://nodejs.org/"
            )
            return False

        # Verificar npm
        if not shutil.which("npm"):
            print_error("npm no está instalado")
            return False

        # Verificar versiones
        try:
            node_version = subprocess.run(
                ["node", "--version"], capture_output=True, text=True
            ).stdout.strip()
            npm_version = subprocess.run(
                ["npm", "--version"], capture_output=True, text=True
            ).stdout.strip()

            print_success(f"Node.js encontrado: {node_version}")
            print_success(f"npm encontrado: {npm_version}")

            # Verificar versión mínima de Node.js
            node_major = int(node_version[1:].split(".")[0])
            if node_major < 18:
                print_error(
                    f"Se requiere Node.js 18 o superior. Versión actual: {node_version}"
                )
                return False

        except Exception as e:
            print_error(f"Error al verificar versiones: {e}")
            return False

        print_success("Todas las dependencias del sistema están disponibles")
        return True

    def setup_environment(self) -> bool:
        """Configura el entorno del proyecto."""
        print_header("CONFIGURANDO ENTORNO")

        # Verificar si estamos en el directorio correcto
        if not self.package_json_path.exists():
            print_error(
                "No se encontró package.json. Asegúrate de estar en el directorio Frontend/"
            )
            return False

        # Verificar si es el proyecto correcto
        try:
            with open(self.package_json_path, "r") as f:
                package_data = json.load(f)
                if package_data.get("name") != "sheily-landing-next":
                    print_error("Este no parece ser el proyecto Sheily AI Frontend")
                    return False
        except Exception as e:
            print_error(f"Error al leer package.json: {e}")
            return False

        print_success("Directorio del proyecto verificado")

        # Crear archivo .env.local si no existe
        if not self.env_local_path.exists():
            print_status("Creando archivo .env.local...")
            env_content = f"""# Configuración del Frontend Sheily AI
NODE_ENV=development
PORT={self.port}
HOSTNAME={self.hostname}
NEXTAUTH_SECRET=sheily_ai_frontend_secret_key_development_{int(time.time())}
BACKEND_URL=http://localhost:8000
NEXTAUTH_URL=http://{self.hostname}:{self.port}
"""
            try:
                with open(self.env_local_path, "w") as f:
                    f.write(env_content)
                print_success("Archivo .env.local creado")
            except Exception as e:
                print_error(f"Error al crear .env.local: {e}")
                return False
        else:
            print_success("Archivo .env.local ya existe")

        # Configurar variables de entorno
        os.environ["NODE_ENV"] = "development"
        os.environ["PORT"] = str(self.port)
        os.environ["HOSTNAME"] = self.hostname

        print_success("Entorno configurado correctamente")
        return True

    def check_node_dependencies(self) -> bool:
        """Verifica las dependencias de Node.js."""
        print_header("VERIFICANDO DEPENDENCIAS DE NODE.JS")

        if not self.node_modules_path.exists():
            print_warning("node_modules no encontrado. Instalando dependencias...")
            return self.install_dependencies()

        print_success("node_modules encontrado")

        # Verificar dependencias
        try:
            result = subprocess.run(
                ["npm", "ls", "--depth=0"],
                capture_output=True,
                text=True,
                cwd=self.project_dir,
            )
            if (
                "UNMET DEPENDENCY" in result.stdout
                or "UNMET DEPENDENCY" in result.stderr
            ):
                print_warning("Dependencias faltantes detectadas. Reinstalando...")
                return self.install_dependencies()
            else:
                print_success("Todas las dependencias están instaladas correctamente")
                return True
        except Exception as e:
            print_warning(f"Error al verificar dependencias: {e}")
            return self.install_dependencies()

    def install_dependencies(self) -> bool:
        """Instala las dependencias de Node.js."""
        print_header("INSTALANDO DEPENDENCIAS")

        try:
            # Limpiar instalación anterior si existe
            if self.node_modules_path.exists():
                print_status("Limpiando instalación anterior...")
                shutil.rmtree(self.node_modules_path)
                package_lock = self.project_dir / "package-lock.json"
                if package_lock.exists():
                    package_lock.unlink()

            # Limpiar caché de npm
            print_status("Limpiando caché de npm...")
            subprocess.run(
                ["npm", "cache", "clean", "--force"], cwd=self.project_dir, check=True
            )

            # Instalar dependencias
            print_status("Instalando dependencias...")
            subprocess.run(["npm", "install"], cwd=self.project_dir, check=True)

            print_success("Dependencias instaladas correctamente")
            return True

        except subprocess.CalledProcessError as e:
            print_error(f"Error al instalar dependencias: {e}")
            return False
        except Exception as e:
            print_error(f"Error inesperado: {e}")
            return False

    def check_nextjs_config(self) -> bool:
        """Verifica la configuración de Next.js."""
        print_header("VERIFICANDO CONFIGURACIÓN DE NEXT.JS")

        config_files = [
            "next.config.cjs",
            "tsconfig.json",
            "tailwind.config.ts",
            "postcss.config.cjs",
        ]

        for config_file in config_files:
            config_path = self.project_dir / config_file
            if config_path.exists():
                print_success(f"{config_file} encontrado")
            else:
                print_error(f"{config_file} no encontrado")
                return False

        # Verificar configuración de TypeScript
        try:
            print_status("Verificando configuración de TypeScript...")
            subprocess.run(
                ["npx", "tsc", "--noEmit", "--skipLibCheck"],
                cwd=self.project_dir,
                check=False,
            )
            print_success("Configuración de TypeScript verificada")
        except Exception as e:
            print_warning(f"Errores de TypeScript detectados, pero continuando...")

        print_success("Configuración de Next.js verificada")
        return True

    def check_audio_files(self) -> bool:
        """Verifica los archivos de audio."""
        print_header("VERIFICANDO ARCHIVOS DE AUDIO")

        audio_dir = self.project_dir / "public" / "sounds"
        required_files = [
            "062708_laser-charging-81968.mp3",
            "whoosh-drum-hits-169007.mp3",
        ]

        if not audio_dir.exists():
            print_warning("Directorio de sonidos no encontrado. Creando...")
            audio_dir.mkdir(parents=True, exist_ok=True)

        missing_files = []
        for audio_file in required_files:
            if not (audio_dir / audio_file).exists():
                missing_files.append(audio_file)

        if missing_files:
            print_warning(f"Archivos de audio faltantes: {', '.join(missing_files)}")
            print_status("Los sonidos se generarán sintéticamente como fallback")
        else:
            print_success("Todos los archivos de audio están disponibles")

        return True

    def check_ports(self) -> bool:
        """Verifica la disponibilidad de puertos."""
        print_header("VERIFICANDO PUERTOS")

        # Verificar puerto 3000
        try:
            if platform.system() == "Windows":
                cmd = f"netstat -ano | findstr :{self.port}"
                result = subprocess.run(cmd, shell=True, capture_output=True, text=True)
                if result.stdout.strip():
                    print_warning(f"Puerto {self.port} está en uso")
                    print_status("Intentando liberar el puerto...")
                    self.cleanup_previous_processes()
            else:
                cmd = f"lsof -ti:{self.port}"
                result = subprocess.run(cmd, shell=True, capture_output=True, text=True)
                if result.stdout.strip():
                    print_warning(f"Puerto {self.port} está en uso")
                    print_status("Intentando liberar el puerto...")
                    self.cleanup_previous_processes()
        except Exception as e:
            print_warning(f"No se pudo verificar el puerto: {e}")

        print_success(f"Puerto {self.port} está disponible")

        # Verificar puerto 8000 (backend)
        try:
            if platform.system() == "Windows":
                cmd = "netstat -ano | findstr :8000"
                result = subprocess.run(cmd, shell=True, capture_output=True, text=True)
                if result.stdout.strip():
                    print_success("Backend detectado en puerto 8000")
                else:
                    print_warning("Backend no detectado en puerto 8000")
            else:
                cmd = "lsof -ti:8000"
                result = subprocess.run(cmd, shell=True, capture_output=True, text=True)
                if result.stdout.strip():
                    print_success("Backend detectado en puerto 8000")
                else:
                    print_warning("Backend no detectado en puerto 8000")
        except Exception as e:
            print_warning(f"No se pudo verificar el backend: {e}")

        print_status("El frontend funcionará en modo standalone si es necesario")
        return True

    def start_frontend(self) -> None:
        """Inicia el frontend."""
        print_header("INICIANDO FRONTEND SHEILY AI")

        print_status(f"Iniciando servidor de desarrollo en puerto {self.port}...")
        print_status(f"URL: http://{self.hostname}:{self.port}")
        print_status("Presiona Ctrl+C para detener el servidor")

        try:
            # Iniciar el servidor
            env = os.environ.copy()
            env.update(
                {
                    "NODE_ENV": "development",
                    "PORT": str(self.port),
                    "HOSTNAME": self.hostname,
                }
            )

            subprocess.run(
                ["npm", "run", "dev"], cwd=self.project_dir, env=env, check=True
            )

        except subprocess.CalledProcessError as e:
            print_error(f"Error al iniciar el frontend: {e}")
        except KeyboardInterrupt:
            print_status("Servidor detenido por el usuario")
        except Exception as e:
            print_error(f"Error inesperado: {e}")

    def run(self) -> bool:
        """Ejecuta el proceso completo de arranque."""
        print_header("ARRANQUE DEL FRONTEND SHEILY AI")
        print_status("Iniciando proceso de arranque completo...")

        try:
            # Ejecutar todas las verificaciones y configuraciones
            if not self.check_system_dependencies():
                return False

            if not self.setup_environment():
                return False

            if not self.check_node_dependencies():
                return False

            if not self.check_nextjs_config():
                return False

            if not self.check_audio_files():
                return False

            if not self.check_ports():
                return False

            print_header("VERIFICACIONES COMPLETADAS")
            print_success("✅ Sistema de dependencias verificado")
            print_success("✅ Entorno configurado")
            print_success("✅ Configuración de Next.js validada")
            print_success("✅ Puertos verificados")
            print_success("✅ Archivos de audio verificados")

            print_status("Iniciando frontend...")
            self.start_frontend()
            return True

        except Exception as e:
            print_error(f"Error durante el arranque: {e}")
            return False


def signal_handler(signum, frame):
    """Maneja las señales de interrupción."""
    print_status("Recibida señal de interrupción. Limpiando...")
    sys.exit(0)


def main():
    """Función principal."""
    # Configurar manejador de señales
    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)

    # Crear y ejecutar el starter
    starter = FrontendStarter()
    success = starter.run()

    if not success:
        print_error("El arranque del frontend falló")
        sys.exit(1)


if __name__ == "__main__":
    main()
